lumicks/pylake/kymotracker/detail/peakfinding.py

```diff
@@ -75,13 +75,143 @@ def flatten(self):
+    bias, we define a virtual window that's symmetric around the spot position. In practice, this
+    involves removing fractional pixels at the edges. If we want to shift the center of the image
+    by 1 pixel, we need to move an edge point 2 pixels.
+
+    Fractional pixels are dealt with by down-weighting edge pixels and recomputing a new pixel
+    center for those pixels.
+
+    For more information on this method, please refer to [1]_.
+
+    References
+    ----------
+    .. [1] Berglund, A. J., McMahon, M. D., McClelland, J. J., & Liddle, J. A. (2008).
+           Fast, bias-free algorithm for tracking single particles with variable size and
+           shape. Optics express, 16(18), 14064-14075.
+
+    Parameters
+    ----------
+    position : np.ndarray
+        Subpixel coordinates obtained with regular centroid method.
+    data : np.ndarray
+        2D array
+    tolerance : float
+        Convergence tolerance
+    max_iterations : int
+        Maximum iterations
+    epsilon : float
+        Small value to make sure we don't divide by zero
+    """
+    lb = np.tile(0.0, data.shape[0])
+    ub = np.tile(float(data.shape[1]), data.shape[0])
+    index_array = np.tile(np.arange(data.shape[1]), (data.shape[0], 1))
+
+    # In this function we define the internal coordinate system with the origin at the pixel edge
+    last_position = position + 0.5
+    for _ in range(max_iterations):
+        selection, centers, weights = bounds_to_centroid_data(index_array, lb, ub)
+
+        range_center = (lb + ub) / 2.0
+        weights *= data
+        weighted_coord = np.sum(weights * centers, axis=1)
+        weighted_sum = np.sum(weights, axis=1)
+        empty_pixels = weighted_sum < epsilon
+        weighted_sum[empty_pixels] = 1.0  # We are not using these anyway
+
+        est_position = weighted_coord / weighted_sum
+
+        # Only consider those pixels where we had sufficient data
+        sufficient_data = np.logical_and(ub - lb > 2, np.logical_not(empty_pixels))
+        position[sufficient_data] = est_position[sufficient_data]
+
+        move_lb = position > range_center
+        lb[move_lb] = lb[move_lb] + 2.0 * (position[move_lb] - range_center[move_lb])
+        move_ub = np.logical_not(move_lb)
+        ub[move_ub] = ub[move_ub] + 2.0 * (position[move_ub] - range_center[move_ub])
+
+        if np.max(np.abs(position - last_position)) < tolerance:
+            break
+
+        # We change position in place each iteration, so we need to be explicit about the copy
+        last_position = position.copy()
+
+    # Internally, we used a coordinate system that has pixel centers at 0.5, 1.5 etc. This is
+    # practical because it simplifies the code. Externally, we define pixel centers at 0, 1, 2 etc,
+    # so we subtract a half here.
+    return position - 0.5
+
+
+def _clip_kernel_to_edge(max_half_width, coordinates, dataset_width):
+    """This function finds the number of pixels between the coordinates and the image edge.
```

Comment:
```suggestion
    """Find the number of pixels between the coordinates and the image edge.
```

from https://peps.python.org/pep-0257/

> The docstring is a phrase ending in a period. It prescribes the function or method’s effect as a command (“Do this”, “Return that”), not as a description; e.g. don’t write “Returns the pathname …”.

Reply:
Ah yes. Bad habit. Thanks for pointing it out. Fixing!